internal/repository/indexmap.go

```diff
@@ -0,0 +1,226 @@
+				return
+			}
+			e = e.next
+		}
+	}
+}
+
+// foreachWithID calls fn for all entries with the given id.
+func (m *indexMap) foreachWithID(id restic.ID, fn func(*indexEntry)) {
+	if len(m.buckets) == 0 {
+		return
+	}
+
+	h := m.hash(id)
+	for e := m.buckets[h]; e != nil; e = e.next {
+		if e.id != id {
+			continue
+		}
+		fn(e)
+	}
+}
+
+// get returns the first entry for the given id.
+func (m *indexMap) get(id restic.ID) *indexEntry {
+	if len(m.buckets) == 0 {
+		return nil
+	}
+
+	h := m.hash(id)
+	for e := m.buckets[h]; e != nil; e = e.next {
+		if e.id == id {
+			return e
+		}
+	}
+	return nil
+}
+
+var cpus = runtime.GOMAXPROCS(0)
+
+func (m *indexMap) grow() {
+	old := m.buckets
+	m.buckets = make([]*indexEntry, growthFactor*len(m.buckets))
+
+	const maxSequential = 512
+	if len(old) <= maxSequential {
+		for _, e := range old {
+			m.take(e)
+		}
+		return
+	}
+
+	m.wg.Add(cpus)
+	for cpu := 0; cpu < cpus; cpu++ {
+		go func(cpu int) {
+			for i := cpu; i < len(old); i += cpus {
+				m.take(old[i])
+			}
+			m.wg.Done()
+		}(cpu)
+	}
+
+	m.wg.Wait()
+	m.wg = sync.WaitGroup{} // Reset for reuse.
+}
+
+func (m *indexMap) hash(id restic.ID) uint {
+	// We use siphash with a randomly generated 128-bit key, to prevent
+	// backups of specially crafted inputs from degrading performance.
+	// While SHA-256 should be collision-resistant, but for hash table indices
+	// we use only a few bits of it and findind collisions for those is
+	// much easier than breaking the whole algorithm.
+	h := uint(siphash.Hash(m.key0, m.key1, id[:]))
+	return h & uint(len(m.buckets)-1)
+}
+
+func (m *indexMap) init() {
+	const initialBuckets = 64
+	m.buckets = make([]*indexEntry, initialBuckets)
+
+	var buf [16]byte
+	if _, err := rand.Read(buf[:]); err != nil {
+		panic(err) // Very little we can do here.
+	}
+	m.key0 = binary.LittleEndian.Uint64(buf[:8])
+	m.key1 = binary.LittleEndian.Uint64(buf[8:])
+}
+
+func (m *indexMap) len() uint { return m.numentries }
+
+func (m *indexMap) newEntry() *indexEntry {
+	// Allocating in batches means that we get closer to optimal space usage,
+	// as Go's malloc will overallocate for structures of size 56 (indexEntry
+	// on amd64).
+	const entryAllocBatch = 8
+
+	if m.free == nil {
+		free := new([entryAllocBatch]indexEntry)
+		for i := range free[:len(free)-1] {
+			free[i].next = &free[i+1]
+		}
+		m.free = &free[0]
+	}
+
+	e := m.free
+	m.free = m.free.next
+
+	return e
+}
+
+// Take transfers e and its entire chain to m.
+func (m *indexMap) take(e *indexEntry) {
+	for e != nil {
+		h := m.hash(e.id)
+		next := e.next
+		e.next = m.buckets[h]
+		m.buckets[h] = e
+		e = next
+	}
+}
+
+// Take transfers e and its entire chain to m.
+func (m *indexMap) moveto(m2 *indexMap, fn func(e *indexEntry)) {
+	if m2.numentries == 0 { // Lazy initialization.
+		m2.buckets = make([]*indexEntry, 64)
+	}
+
+	m2.numentries += m.numentries
+	// grow m2 if needed
+	for m2.numentries > maxLoad*uint(len(m2.buckets)) {
+		m2.grow()
```

Comment:
This reallocates in a loop. If grow would take the desired number of elements as an argument, it could then round that up to a power of two using

```go
if n&(n-1) != 0 {
	n = 1 << bits.Len(n)
}
```

Reply:
I moved the loop into `grow()`. That should not be too expensive..